### PR TITLE
[libc++] Remove the allocator<const T> extension

### LINKED ISSUE #73665: [libc++] Remove allocator<T const> extension
See https://reviews.llvm.org/D120996 and a54d028895c91da356a4aaf30e27a5a5b90dd313. Also https://github.com/llvm/llvm-project/issues/73664.

### PR DESCRIPTION
**Pull request by ldionne**
This effort has quite a history:
- This was first attempted in 2022 via bed3240bf7d1, which broke std::shared_ptr<T const> and caused the change to be reverted in 9138666f5464.
- We then re-attempted landing the change in 276ca87382b8 after fixing std::shared_ptr, but reports were made that this broke code en masse within Google. This led to the patch being reverted again in a54d028895c9 with the goal to land this again with a migration path for vendors.

This patch re-lands the removal while providing a migration path for vendors by providing the `_LIBCPP_ENABLE_REMOVED_ALLOCATOR_CONST` macro. This macro will be honored for the LLVM 19 release and will be removed after that, at which point allocator<const T> will be removed unconditionally.

Fixes #73665

### CODE REVIEW

**Comment by ldionne** [comment]
Pinging @llvm/libcxx-vendors for awareness.

**Comment by zibi2** [comment]
We are fine with this patch. Thank you for giving us grace period.

File: libcxx/docs/ReleaseNotes/19.rst [hunk 1]
```diff
@@ -121,6 +121,12 @@ Deprecations and Removals
 - The ``_LIBCPP_DISABLE_ADDITIONAL_DIAGNOSTICS`` macro has been removed and is not honored anymore. Additional
   warnings provided by libc++ as a matter of QoI will now be provided unconditionally.
 
+- libc++ no longer supports ``std::allocator<const T>`` and containers of ``const``-qualified element type, such
+  as ``std::vector<const T>`` and ``std::list<const T>``. This used to be supported as an undocumented extension.
+  If you were using ``std::vector<const T>``, replace it with ``std::vector<T>`` instead. The
+  ``_LIBCPP_ENABLE_REMOVED_ALLOCATOR_CONST`` macro can be defined to temporarily re-enable this extension as
+  folks transition their code. This macro will be honored for one released and ignored starting in LLVM 20.
+
```
Comment (cjdb) [inline libcxx/docs/ReleaseNotes/19.rst:129]:
```suggestion
- libc++ no longer supports ``std::allocator<const T>`` and containers of ``const``-qualified element type, such
  as ``std::vector<const T>`` and ``std::list<const T>``. This used to be supported as an undocumented extension.
  If you were using ``std::vector<const T>``, replace it with ``std::vector<T>`` instead. The
  ``_LIBCPP_ENABLE_REMOVED_ALLOCATOR_CONST`` macro can be defined to temporarily re-enable this extension as
  folks transition their code. This macro will be honored for one released and ignored starting in LLVM 20.
  To assist with the clean-up process, consider running your code through Clang Tidy, with [std-allocator-const](https://clang.llvm.org/extra/clang-tidy/checks/portability/std-allocator-const.html) enabled.
```

**Review by cjdb — APPROVED** [review]
Thank you for providing the opt-out mechanism!

**Pull request merged**